### PR TITLE
Move Static Dake Model Calculation to Connection Class

### PR DESCRIPTION
**Pull request by bska**
Certain parts of the Dake model D-factor correlation relation depend only on static parameters at the cell level.  Calculate the static product once, and store it as a member of `Connection::CTFProperties` (PR #3763). In particular, the new member
```C++
double CTFProperties::static_dfac_corr_coeff
```
stores the product
```math
A \cdot (K/K_0)^B \cdot \phi^C \cdot K_e/(h\cdot r_w)
```
with `A`, `B`, and `C` being the Dake model parameters.  We compute new values for the `static_dfac_corr_coeff` as part of processing the `WDFACCOR` keyword. This static product will also be output to the restart file in follow-up PRs, and will be used to simplify evaluating the full dynamic Dake model correlation.  Since we need the porosity value $\phi$, we also make this available through the `ScheduleGrid` structure.

We finally switch to calculating the D-factor in terms of the `Connection` class's `CTFProperties`.  In particular, we switch the API of `WDFAC::getDFactor()` to take the dynamic viscosity and gas density values in addition to a `Connection` object.  We then reimplement the Dake model correlation in terms of the `Connection`'s `static_dfac_corr_coeff` since this bypasses the exponentiation and, in the future, will simplify restarting a simulation run using the `WDFACCOR` keyword.

As a means to this end, we also pull the Dake model parameter out to a new helper structure,
```C++
struct WDFAC::Correlation
```
as these parameters must also be output to the restart file.  Making it easier to refer to the parameters as a collection will simplify that restart file writing code.

### CODE REVIEW

**Comment by bska** [comment]
I am creating this PR in draft mode for two reasons

1. It depends on, and contains, the upstream PR #3763.
2. It changes the `WDFAC::getDFactor()` API and must therefore be accompanied by a downstream PR to match that API change.

I will keep the PR in draft state until it is ready for revew.

**Comment by bska** [comment]
The upstream PR has been merged into the master branch.  Marking this PR as "ready for review" now.

**Comment by bska** [comment]
jenkins build this opm-simulators=4992 please

**Comment by bska** [comment]
jenkins build this opm-simulators=4992 please

File: src/opm/input/eclipse/Schedule/Well/WellConnections.cpp [hunk 1]
```diff
@@ -492,6 +509,9 @@ The cell ({},{},{}) in well {} is not active and the connection will be ignored)
             // PolymerMW module.
             ctf_props.re = std::sqrt(D[0] * D[1] / angle * 2);
 
+            ctf_props.static_dfac_corr_coeff =
+                staticForchheimerCoeffcient(ctf_props, props->poro, wdfac);
```
Comment (akva2) [inline src/opm/input/eclipse/Schedule/Well/WellConnections.cpp:513]: Coefficient

Comment (bska) [reply src/opm/input/eclipse/Schedule/Well/WellConnections.cpp]: > Coefficient

Good catch.  I've pushed an update to address the misprint.

File: opm/input/eclipse/Schedule/Well/WDFAC.hpp [hunk 2]
```diff
@@ -137,6 +135,52 @@ namespace Opm {
...
+                return this->scaledWellLevelDFactor(this->m_d, conn);
+
+            case WDFacType::DAKEMODEL:
+                return this->dakeModelDFactor(rhoGS(), gas_visc(), conn);
+
+            case WDFacType::CON_DFACTOR:
+                return this->connectionLevelDFactor(conn);
+            }
+
+            return 0.0;
```
[the remark this answers is not in the document]
Comment (bska) [reply opm/input/eclipse/Schedule/Well/WDFAC.hpp]: > assert?

Maybe.  I've pushed an update which throws an exception instead.  For what it's worth, the current master version also [returns zero](https://github.com/OPM/opm-common/blob/4485bdfa83f61f4663d5291365b24d91eb387304/src/opm/input/eclipse/Schedule/Well/WDFAC.cpp#L132) in this case.

**Comment by bska** [comment]
jenkins build this opm-simulators=4992 please

**Comment by bska** [comment]
jenkins build this opm-simulators=4992 please

**Comment by bska** [comment]
jenkins build this opm-simulators=4992 please

**Comment by totto82** [comment]
Thanks for the refactoring and the clean-ups. Jenkins is happy and so am I. Will merge.

**Pull request merged**